File: 알고리즘 스터디 문제들/Part 5 - 다이나믹 프로그래밍/문제/1, 2, 3 더하기.cpp
#include <iostream>
#include <array>
using namespace std;

array<int, 11 + 1> memo = {};

int main()
{
    memo[0] = 1;
    for (int i = 1; i <= 11; i++)
    {
        if (i - 3 >= 0)
            memo[i] += memo[i - 3];
        if (i - 2 >= 0)
            memo[i] += memo[i - 2];
        if (i - 1 >= 0)
            memo[i] += memo[i - 1];
    }

    size_t T, N;
    cin >> T;
    for (size_t i = 0; i < T; i++)
    {
        cin >> N;
        cout << memo[N] << '\n';
    }
}

File: 알고리즘 스터디 문제들/Part 5 - 다이나믹 프로그래밍/문제/1, 2, 3 더하기.py
memo = [0 for i in range(12)]
memo[1] = 1
memo[2] = memo[1] + 1
memo[3] = memo[2] + memo[1] + 1

for i in range(4, 12):
    memo[i] = memo[i - 1] + memo[i - 2] + memo[i - 3]

T = int(input())
for i in range(T):
    N = int(input())
    print(memo[N])
